cpumanager: validate topology in static policy

ipuustin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This patch adds a check for the static policy state validation. The check fails if the CPU topology obtained from cadvisor doesn't match with the current topology in the state file.

If the CPU topology has changed in a node, cpumanager static policy might try to assign non-present cores to containers.

For example in my test case, static policy had the default CPU set of `0-1,4-7`. Then kubelet was shut down and CPU 7 was offlined. After restarting the kubelet, CPU manager tries to assign the non-existent CPU 7 to containers which don't have exclusive allocations assigned to them:

    Error response from daemon: Requested CPUs are not available - requested 0-1,4-7, available: 0-6)

This breaks the exclusivity, since the CPUs from the shared pool don't get assigned to non-exclusive containers, meaning that they can execute on the exclusive CPUs.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Added CPU Manager state validation in case of changed CPU topology.
```

ipuustin committed:
cpumanager: validate topology in static policy.

This patch adds a check for the static policy state validation. The
check fails if the CPU topology obtained from cadvisor doesn't match
with the current topology in the state file.

If the CPU topology has changed in a node, cpu manager static policy
might try to assign non-present cores to containers.

For example in my test case, static policy had the default CPU set of
0-1,4-7. Then kubelet was shut down and CPU 7 was offlined. After
restarting the kubelet, CPU manager tries to assign the non-existent CPU
7 to containers which don't have exclusive allocations assigned to them:

 Error response from daemon: Requested CPUs are not available - requested 0-1,4-7, available: 0-6)

This breaks the exclusivity, since the CPUs from the shared pool don't
get assigned to non-exclusive containers, meaning that they can execute
on the exclusive CPUs.

neolit123 commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -145,6 +145,20 @@ func (p *staticPolicy) validateState(s state.State) error {
+		totalKnownCPUs = totalKnownCPUs.Union(cset)
+	}
+	if !totalKnownCPUs.Equals(p.topology.CPUDetails.CPUs()) {
+		return fmt.Errorf("current CPU topology (%s) doesn't match with CPUs in state (%s)",
```
"...the CPUs stored in the state..."

ConnorDoyle replied:
nit: this has less to do with topology than e.g. "the set of available CPUs". I think adding topology in this error message might confuse a user. nit2: could we format the cpuset so it's surrounded by double-quotes instead of parens? That would match the other logs in this module better IIRC.

neolit123 reviewed (commented):
/sig node
/kind bug

neolit123 commented:
@ipuustin thanks, probably needs a release note.

dims commented:
/uncc

ConnorDoyle commented:
/assign

ConnorDoyle reviewed (changes requested):
Other than log message & format, this LGTM. Ping when updated to get approved.

ConnorDoyle commented:
@ipuustin mind adding a note to the [docs](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/) about:

1. The CPU manager doesn't support off/on-lining CPUs at runtime.
2. Warn about this scenario, e.g. if the set of online CPUs changes then you need to perform a node drain and manually reset the CPU manager state by deleting the file.

ConnorDoyle commented:
/ok-to-test
/retest

ConnorDoyle commented on pkg/kubelet/cm/cpumanager/policy_static.go:
```diff
@@ -145,6 +145,23 @@ func (p *staticPolicy) validateState(s state.State) error {
+		totalKnownCPUs = totalKnownCPUs.Union(cset)
+	}
+	if !totalKnownCPUs.Equals(p.topology.CPUDetails.CPUs()) {
+		return fmt.Errorf("current set of available CPUs \"%s\" doesn't match with CPUs in state \"%s\"",
```
Could you please add a unit test for each case of this conditional? It should be possible to either extend the existing test for Start or create a new test for this validateState function.

ipuustin replied:
I added two test cases now. The test where the amount of CPUs equals in state and topology was already present ("non-corrupted state"), so I just tested the cases where the number of CPUs doesn't match.

ipuustin committed:
cpumanager: add test for available CPUs in static policy.

Test the cases where the number of CPUs available in the system is
smaller or larger than the number of CPUs known in the state, which
should lead to a panic. This covers both CPU onlining and offlining. The
case where the number of CPUs matches is already covered by the
"non-corrupted state" test.

ipuustin commented:
/retest

ConnorDoyle reviewed (approved):
:100:
/approve
/lgtm
